File: medium/49.py
import collections
from typing import List

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        ans = collections.defaultdict(list)

        def get_key(arr: List[int]) -> str:
            full_str = "" 
            for num in arr:
                full_str += f"{num}#"
            return full_str


        for s in strs:
            count = [0] * 26

            for c in s:
                count[ord(c) - ord('a')] += 1

            ans[get_key(count)].append(s)

        return ans.values()

File: medium/49.cpp
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> store;
        
        for(auto& word: strs) {
          auto counter = getLetterCount(word);
          string hash = getHash(counter);
          store[hash].emplace_back(word);
        }


        vector<vector<string>> result;
        for(auto& kv: store) {
          result.emplace_back(kv.second);
        }

        return result;
    }
private:
    vector<int> getLetterCount(string word) {
      int NO_ASCII_LETTERS = 26;
      vector<int> counter(NO_ASCII_LETTERS);

      for(char& letter: word) {
        counter[letter]++;
      }

      return counter;
    }

    string getHash(vector<int>& arr) {
      string hash = "";

      for(int& num: arr) {
        hash.append(to_string(num) + '#'); 
      }
      return hash;
    }

};
